### PR TITLE
Add vector_similarity scalar function (euclidean based)

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Relates to https://github.com/crate/crate/issues/14801

### CODE REVIEW

**Comment by BaurzhanSakhariev** [comment]
Putting to draft as may add more types (cosine, dot_product) as a scalar argument.

**Comment by BaurzhanSakhariev** [comment]
Checking the implementation, I think vectore_similarity_score would be a better name + forgot to check same length for vectors https://github.com/apache/lucene/pull/12253/files#diff-55fac07d1f7553c8e87e2841eeb008596d654388f05a4a86133da54076bca6e8R45

**Comment by BaurzhanSakhariev** [comment]
Some databases expose actual distance (opened https://github.com/crate/crate/issues/15835 to track that as it's also a valid use case).

Here we return score, a value from 0 to 1. 

This is also aligned with what we use in `knn_match`/`float_vector` type (taking to account that _score of the knn_match can be affected by something else, hence dedicated scalar).

File: docs/general/builtins/scalar-functions.rst [hunk 1]
```diff
@@ -4438,6 +4438,29 @@ An example::
     SELECT 2 rows in set (... sec)
 
 
+.. _scalar_vector_similarity:
+
+
+``vector_similarity(float_vector, float_vector, float)``
```
Comment (seut) [inline docs/general/builtins/scalar-functions.rst:4444]:
```suggestion
``vector_similarity(float_vector, float_vector)``
```

Comment (seut) [inline docs/general/builtins/scalar-functions.rst:4444]: I think this scalar should be placed under a new `vector` section.

File: docs/general/builtins/scalar-functions.rst [hunk 2]
```diff
@@ -4438,6 +4438,29 @@ An example::
...
+--------------------------------------------------------
+
+Returns similarity of 2 :ref:`FLOAT_VECTORS <type-float_vector>`.
```
Comment (seut) [inline docs/general/builtins/scalar-functions.rst:4447]: We should clarify on the return data type.
```suggestion
Returns similarity of 2 :ref:`FLOAT_VECTORS <type-float_vector>` as a :ref:`FLOAT <type-float>` typed value.
```

File: server/src/main/java/io/crate/expression/scalar/VectorSimilarityFunction.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    public static void register(Functions.Builder module) {
+        module.add(
+            Signature.scalar(
+                "vector_similarity",
+                FloatVectorType.INSTANCE_ONE.getTypeSignature(),
+                FloatVectorType.INSTANCE_ONE.getTypeSignature(),
+                DataTypes.FLOAT.getTypeSignature()
+            ).withFeature(Scalar.Feature.NULLABLE),
+            VectorSimilarityFunction::new
```
Comment (seut) [inline server/src/main/java/io/crate/expression/scalar/VectorSimilarityFunction.java:44]: Can't we use `BinaryScalar` instead?

File: server/src/main/java/io/crate/types/FloatVectorType.java [hunk 4]
```diff
@@ -166,6 +166,9 @@ public long valueBytes(float @Nullable [] value) {
 
     @Override
     public TypeSignature getTypeSignature() {
+        if (dimensions == 1) {
+            return new TypeSignature(NAME);
+        }
```
Comment (seut) [inline server/src/main/java/io/crate/types/FloatVectorType.java:171]: Why is this needed?

File: server/src/main/java/io/crate/expression/scalar/VectorSimilarityFunction.java (hunk 3 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/expression/scalar/VectorSimilarityFunction.java:44]: No, because Input/Return types are different.

File: docs/general/builtins/scalar-functions.rst (hunk 1 above)
Comment (BaurzhanSakhariev) [reply docs/general/builtins/scalar-functions.rst]: Done in https://github.com/crate/crate/pull/15832/commits/2f5dddf244b28321fd2b792ba1cd59fc39ab99fd.

Should I move `knn_match` there as well? It's currently in the "special functions" section.

File: server/src/main/java/io/crate/types/FloatVectorType.java (hunk 4 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/types/FloatVectorType.java:171]: Function is registered with fixed vector dimension 1. 
It's not enough to include the type into `ALLOW_BASENAME_MATCH`. 

Without this change scalar function call fails with 
` The signature type of the based data type parameters can only be: INTEGER_LITERAL_SIGNATURE but was TYPE_SIGNATURE`

I believe it happens because FloatVectorType.getTypeSignature creates `IntegerLiteralTypeSignature` via `TypeSignature.of(dimensions)` and `SignatureBinder.applyBoundVariables` loses that information [here](https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/metadata/functions/SignatureBinder.java#L180).

I removed this check and instead tried adding 
if (typeSignature instanceof IntegerLiteralTypeSignature) {
            return typeSignature;
}

to `SignatureBinder.applyBoundVariables` and tests passed.

Looks like fixing `applyBoundVariables` would be a better option - and we will be able to remove that check also in all BitType/CharacterType/StringType?

See also https://github.com/crate/crate/pull/13278/commits/abc3f7184080049960395960401cb7e93af02659

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/types/FloatVectorType.java]: I merged linked PR and removed this "hack"

**Review by seut — APPROVED** [review]
